### ISSUE TITLE
doc improvement: clarify intended usage and level of support

### ISSUE THREAD

**Issue by eeaton** [terraform-google-modules/terraform-example-foundation#1239]
### TL;DR

This repo is intended to be forked into the user's own environment, then customized and maintained in their own version control system. We do not intend to support use cases that use this repo as a remote reference. However, this has not been clearly documented.

I will add documentation throughout the repo to make it more clear about what we intend to support:

- This repo is a public example of code to deploy  resources aligned to best practices in the [Enterprise foundation blueprint]s semantically valid
- Code is semantically valid, pinned to known good versions, and passes terraform validate and lint checks
- The actual resource deployment passes CI tests before merging that confirm the resources can deploy as intended.

And what we do not intend to support: 
- using this repo as a remote reference
- the ability for a customer environment to do in-place upgrades between repo versions after they have customized the code and deployed other resources on top
- support for all versions and customizations that a user might tweak in their own environment

### Terraform Resources

```markdown
n/a
```

### Detailed design

```markdown
n/a
```

### Additional information

https://github.com/terraform-google-modules/terraform-example-foundation/issues/1167 is a good example of why we need to set expectations on the intended usage and level of support.

**Comment by mromascanu123** [comment]
Almost any open source project provides a functional baseline that one can use as is and obviously, if need be, customize.
But in most cases there is no need to customize because the consumers audience of such a OSS distro find the baseline functionality more than sufficient. And no specific support for normal / commonusage is needed except for very special cases.
TEF is an exception and it seems the project is geared at strong development teams rather than end-users. Because as-is a TEF deployment is unusable for real-life use scenarios and is indeed just an example of how to use the TF modules.
This is a major issue for a majority of would-be adopters of the Google Landing Zone – very few (and large) organizations have the capacity to re-develop TEF or engage PSO to do it, and most organizations would simply go with another provider’s turn-key landing zone, be it Azure, AWS or Oracle, which have much lower barriers to adoptions and can be used as-is for most common use scenarios just by tweaking simple configurations (yaml files in case of AWS LZA, vargroups  for Azure).

Maybe need to think **in the short term** to a way of adapting TEF to become usable as-is and customizable if needed, and this via a mechanism of configurability making the code mostly immutable and deployment for most common use scenarios data-driven e.g. via yaml config files just like AWS LZA.

**Comment by eeaton** [comment]
Addressed by PR #1276.

And, followup to the comment from May 16 that I previously overlooked:

We're in agreement that it is unrealistic for this repo to be used exactly as is. Our target audience of highly complex enterprise organizations in regulated industries will inevitably require customization, visibility, and control over many design aspects; we assume that they have a cloud platform team with the skills and responsibility to do this (whether in-house or with partner support). This is a contradictory goal to a turnkey solution.

In cases that want a turnkey solution, you might be interested in [Google Cloud setup](https://cloud.google.com/docs/enterprise/setup-checklist), a guided console-based experience that assumes less prerequisite cloud knowledge and no hands-on coding required. However, by design that tool covers a much simpler scope of features, and does configures foundation features only (no guidance on day 2 operations like guidance on automating through CICD pipelines, modifying the foundation in the future, deploying workloads to the foundation, still assumes that privileged humans create resources directly, etc).

The two tools have different goals and different audiences: as simple as possible but limited scope of control and customization, or some design/engineering effort required but able to address a much more comprehensive scope of control and customization.

I'll close this issue for now, please reopen if you think it needs further discussion.

**Comment by mromascanu123** [comment]
Not sure about the completion criteria having been relied upon to "complete" the fundamental issue raised in my comment.
Unless by technical and business design the TEF distribution is intended to be unusable, as stated in the closure comment.
But instead of burning cycles on esthetics (e.g. renaming non-production to nonproduction) nothing would stop the TEF development team from refactoring the TEF distribution e.g. adding ("as an example" if more appropriate) a basic configurability mechanism replacing the annoying hardcoded "example" code (addressing scheme, regions, etc). This was raised by #1152 , #1153,, #1154, #1155 and #1226.  It would be also more useful, even for an example, to fix the bugs raised in #1228 , #1240 and #1269.
